gadopt/transport_solver.py

```diff
@@ -0,0 +1,372 @@
+
+        # Solver object is set up later to permit editing default solver options.
+        self._solver_ready = False
+
+        self.set_boundary_conditions()
+        self.set_su_nubar()
+        self.set_equation()
+        self.set_solver_options()
+
+    def set_boundary_conditions(self) -> None:
+        """Sets up boundary conditions."""
+        self.strong_bcs = []
+        self.weak_bcs = {}
+
+        for bc_id, bc in self.bcs.items():
+            weak_bc = {}
+
+            for bc_type, value in bc.items():
+                if bc_type == self.strong_bcs_tag:
+                    if self.continuous_solution:
+                        strong_bc = DirichletBC(self.solution_space, value, bc_id)
+                        self.strong_bcs.append(strong_bc)
+                    else:
+                        weak_bc["q"] = value
+                else:
+                    weak_bc[bc_type] = value
+
+            self.weak_bcs[bc_id] = weak_bc
+
+    def set_su_nubar(self) -> None:
+        """Sets up the advection streamline-upwind scheme (Donea & Huerta, 2003).
+
+        Columns of Jacobian J are the vectors that span the quad/hex and can be seen as
+        unit vectors scaled with the dx/dy/dz in that direction (assuming physical
+        coordinates x, y, z aligned with local coordinates).
+        Thus u^T J is (dx * u , dy * v). Following (2.44c), Pe = u^T J / 2 kappa, and
+        beta(Pe) is the xibar vector in (2.44a). Finally, we get the artificial
+        diffusion nubar from (2.49).
+
+        Donea, J., & Huerta, A. (2003).
+        Finite element methods for flow problems.
+        John Wiley & Sons.
+        """
+        if self.su_diffusivity is None:
+            return
+
+        if (u := getattr(self, "u", self.eq_attrs.get("u"))) is None:
+            raise ValueError(
+                "'u' must be included into `eq_attrs` if `su_diffusivity` is given."
+            )
+
+        if not self.continuous_solution:
+            raise TypeError("SU advection requires a continuous function space.")
+
+        log("Using SU advection")
+
+        J = Function(TensorFunctionSpace(self.mesh, "DQ", 1), name="Jacobian")
+        J.interpolate(Jacobian(self.mesh))
+        # Calculate grid Peclet number. Note the use of a lower bound for diffusivity if
+        # a pure advection scenario is considered.
+        Pe = absv(dot(u, J)) / 2 / (self.su_diffusivity + 1e-12)
+        beta_Pe = as_vector([1 / tanh(Pe_i + 1e-6) - 1 / (Pe_i + 1e-6) for Pe_i in Pe])
+        nubar = dot(absv(dot(u, J)), beta_Pe) / 2  # Calculate SU artificial diffusion
+
+        self.eq_attrs["su_nubar"] = nubar
+
+    @abc.abstractmethod
+    def set_equation(self):
+        """Sets up the term contributions in the equation."""
+        raise NotImplementedError
+
+    def set_solver_options(self) -> None:
+        """Sets PETSc solver parameters."""
+        if isinstance(solver_preset := self.solver_parameters, dict):
+            return
```

Comment:
```suggestion
        solver_preset = self.solver_parameters
        if isinstance(solver_preset, dict):
            return
```

Prefer this to be more explicit

Reply:
That is fair. I realise now that I probably copied that logic from [stokes_integrators](https://github.com/g-adopt/g-adopt/blob/3dffa3b1427695f90b313b3f72dfd229808053d7/gadopt/stokes_integrators.py#L241), where it makes sense, but it is not required here. I simplified.